File: public/sdk/emil.js
// EMIL Platform API — minimal JavaScript SDK (Node 18+ / browsers with fetch).
// Usage:
//   import { EmilClient } from 'https://<your-emil-host>/sdk/emil.js'
//   const emil = new EmilClient({ apiKey: 'emil_test_…', baseUrl: 'https://<your-emil-host>' })
//   const { state } = await emil.state()
// Every method returns the parsed JSON body; non-2xx responses throw EmilError
// with .status and .retryAfterSec (honour it on 429).
export class EmilError extends Error {
  constructor(message, status, retryAfterSec) { super(message); this.name = 'EmilError'; this.status = status; this.retryAfterSec = retryAfterSec }
}
export class EmilClient {
  constructor({ apiKey, baseUrl = '', fetchImpl } = {}) {
    if (!apiKey) throw new Error('apiKey is required')
    this.apiKey = apiKey
    this.baseUrl = String(baseUrl).replace(/\/$/, '') + '/api/v1'
    this.fetch = fetchImpl || globalThis.fetch
  }
  async request(method, path, { query, body } = {}) {
    const url = new URL(this.baseUrl + path)
    for (const [k, v] of Object.entries(query || {})) if (v !== undefined && v !== null) url.searchParams.set(k, String(v))
    const res = await this.fetch(url, { method, headers: { 'x-api-key': this.apiKey, ...(body ? { 'Content-Type': 'application/json' } : {}) }, body: body ? JSON.stringify(body) : undefined })
    const text = await res.text()
    let json
    try { json = text ? JSON.parse(text) : {} } catch { json = { raw: text } }
    if (!res.ok) throw new EmilError(json.error || json.message || ('HTTP ' + res.status), res.status, json.retryAfterSec ?? Number(res.headers.get('retry-after') || 0))
    return json
  }
  get(path, query) { return this.request('GET', path, { query }) }
  post(path, body) { return this.request('POST', path, { body }) }
  del(path, query) { return this.request('DELETE', path, { query }) }
  // Account
  ping() { return this.get('/ping') }
  me() { return this.get('/me') }
  usage() { return this.get('/usage') }
  // EMIL
  state() { return this.get('/state') }
  strategies() { return this.get('/strategies') }
  // Market data
  board() { return this.get('/market/board') }
  quotes(symbols) { return this.get('/market/quotes', { symbols: Array.isArray(symbols) ? symbols.join(',') : symbols }) }
  candles(symbol, interval = '1day', outputsize = 90) { return this.get('/market/candles', { symbol, interval, outputsize }) }
  correlation(a, b, days = 365) { return this.get('/market/correlation', { a, b, days }) }
  news(category = 'markets', score = true) { return this.get('/news', { category, score: score ? 1 : 0 }) }
  calendar() { return this.get('/calendar') }
  centralBanks() { return this.get('/calendar/central-banks') }
  report(symbol) { return this.get('/research/report', { symbol }) }
  brief() { return this.get('/research/brief') }
  // Alerts & watchlist
  watchlist() { return this.get('/watchlist') }
  track(symbol) { return this.post('/watchlist', { symbol }) }
  untrack(symbol) { return this.del('/watchlist/' + encodeURIComponent(symbol)) }
  alerts() { return this.get('/alerts') }
  createAlert(symbol, condition, threshold, note) { return this.post('/alerts', { symbol, condition, threshold, note }) }
  deleteAlert(id) { return this.del('/alerts/' + encodeURIComponent(id)) }
  notifications() { return this.get('/notifications') }
  // Journal & portfolio
  journal() { return this.get('/journal') }
  journalWrite(entry) { return this.post('/journal', entry) }
  portfolio() { return this.get('/portfolio') }
  // Paper trading (never live)
  paperVenues() { return this.get('/paper/venues') }
  paperOrders(venue) { return this.get('/paper/orders', { venue }) }
  paperPlace(order) { return this.post('/paper/orders', order) }
  paperCancel(id, venue, symbol) { return this.del('/paper/orders/' + encodeURIComponent(id), { venue, symbol }) }
  // Webhooks
  webhooks() { return this.get('/webhooks') }
  createWebhook(url, events = ['*'], description) { return this.post('/webhooks', { url, events, description }) }
  deleteWebhook(id) { return this.del('/webhooks/' + encodeURIComponent(id)) }
  testWebhook(id) { return this.post('/webhooks/' + encodeURIComponent(id) + '/test') }
  // Bring-your-own data
  ingestQuotes(rows) { return this.post('/ingest/quotes', { rows }) }
  ingestOrders(rows) { return this.post('/ingest/orders', { rows }) }
  ingestPnl(rows) { return this.post('/ingest/pnl', { rows }) }
  ingestSummary() { return this.get('/ingest/summary') }
  // Streaming: browsers' EventSource cannot set headers, so the stream accepts ?key= as well.
  streamUrl(symbols = []) { return this.baseUrl + '/stream?symbols=' + encodeURIComponent(symbols.join(',')) + '&key=' + encodeURIComponent(this.apiKey) }
}
// Verify an incoming EMIL webhook (Node): await verifyWebhook(rawBody, req.headers['x-emil-signature'], secret)
export async function verifyWebhook(rawBody, signatureHeader, secret, toleranceSec = 300) {
  const parts = Object.fromEntries(String(signatureHeader || '').split(',').map((p) => p.split('=')))
  const t = Number(parts.t)
  const v1 = parts.v1
  if (!t || !v1 || Math.abs(Date.now() / 1000 - t) > toleranceSec) return false
  const { createHmac, timingSafeEqual } = await import('node:crypto')
  const expected = createHmac('sha256', secret).update(t + '.' + rawBody).digest('hex')
  return expected.length === v1.length && timingSafeEqual(Buffer.from(expected), Buffer.from(v1))
}

File: public/sdk/emil.py
"""EMIL Platform API - minimal Python SDK (requests).

    from emil import EmilClient
    emil = EmilClient(api_key="emil_test_...", base_url="https://<your-emil-host>")
    print(emil.state())

Non-2xx responses raise EmilError with .status and .retry_after_sec (honour it on 429).
Webhook receivers: verify_webhook(raw_body, signature_header, secret) -> bool
"""
import hashlib
import hmac
import time

import requests


class EmilError(Exception):
    def __init__(self, message, status, retry_after_sec=0):
        super().__init__(message)
        self.status = status
        self.retry_after_sec = retry_after_sec


class EmilClient:
    def __init__(self, api_key, base_url="", timeout=30):
        if not api_key:
            raise ValueError("api_key is required")
        self.api_key = api_key
        self.base_url = base_url.rstrip("/") + "/api/v1"
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers["x-api-key"] = api_key

    def request(self, method, path, params=None, json=None):
        r = self.session.request(method, self.base_url + path, params=params, json=json, timeout=self.timeout)
        try:
            body = r.json() if r.text else {}
        except ValueError:
            body = {"raw": r.text}
        if not r.ok:
            retry = body.get("retryAfterSec") or int(r.headers.get("retry-after", 0) or 0)
            raise EmilError(body.get("error") or body.get("message") or "HTTP %d" % r.status_code, r.status_code, retry)
        return body

    def get(self, path, **params):
        return self.request("GET", path, params={k: v for k, v in params.items() if v is not None} or None)

    def post(self, path, **json):
        return self.request("POST", path, json=json)

    def delete(self, path, **params):
        return self.request("DELETE", path, params={k: v for k, v in params.items() if v is not None} or None)

    # Account / EMIL
    def ping(self): return self.get("/ping")
    def me(self): return self.get("/me")
    def usage(self): return self.get("/usage")
    def state(self): return self.get("/state")
    def strategies(self): return self.get("/strategies")

    # Market data
    def board(self): return self.get("/market/board")
    def quotes(self, symbols):
        return self.get("/market/quotes", symbols=",".join(symbols) if isinstance(symbols, (list, tuple)) else symbols)
    def candles(self, symbol, interval="1day", outputsize=90):
        return self.get("/market/candles", symbol=symbol, interval=interval, outputsize=outputsize)
    def correlation(self, a, b, days=365): return self.get("/market/correlation", a=a, b=b, days=days)
    def news(self, category="markets", score=True): return self.get("/news", category=category, score=1 if score else 0)
    def calendar(self): return self.get("/calendar")
    def central_banks(self): return self.get("/calendar/central-banks")
    def report(self, symbol): return self.get("/research/report", symbol=symbol)
    def brief(self): return self.get("/research/brief")

    # Alerts & watchlist
    def watchlist(self): return self.get("/watchlist")
    def track(self, symbol): return self.post("/watchlist", symbol=symbol)
    def untrack(self, symbol): return self.delete("/watchlist/" + symbol)
    def alerts(self): return self.get("/alerts")
    def create_alert(self, symbol, condition, threshold, note=None):
        return self.post("/alerts", symbol=symbol, condition=condition, threshold=threshold, note=note)
    def delete_alert(self, alert_id): return self.delete("/alerts/" + alert_id)
    def notifications(self): return self.get("/notifications")

    # Journal & portfolio
    def journal(self): return self.get("/journal")
    def journal_write(self, **entry): return self.post("/journal", **entry)
    def portfolio(self): return self.get("/portfolio")

    # Paper trading (never live)
    def paper_venues(self): return self.get("/paper/venues")
    def paper_orders(self, venue=None): return self.get("/paper/orders", venue=venue)
    def paper_place(self, venue, symbol, side, order_type, qty, price=None):
        return self.post("/paper/orders", venue=venue, symbol=symbol, side=side, orderType=order_type, qty=qty, price=price)
    def paper_cancel(self, order_id, venue, symbol=None):
        return self.delete("/paper/orders/" + order_id, venue=venue, symbol=symbol)

    # Webhooks
    def webhooks(self): return self.get("/webhooks")
    def create_webhook(self, url, events=("*",), description=None):
        return self.post("/webhooks", url=url, events=list(events), description=description)
    def delete_webhook(self, endpoint_id): return self.delete("/webhooks/" + endpoint_id)
    def test_webhook(self, endpoint_id): return self.post("/webhooks/" + endpoint_id + "/test")

    # Bring-your-own data
    def ingest_quotes(self, rows): return self.post("/ingest/quotes", rows=rows)
    def ingest_orders(self, rows): return self.post("/ingest/orders", rows=rows)
    def ingest_pnl(self, rows): return self.post("/ingest/pnl", rows=rows)
    def ingest_summary(self): return self.get("/ingest/summary")


def verify_webhook(raw_body, signature_header, secret, tolerance_sec=300):
    parts = dict(p.split("=", 1) for p in (signature_header or "").split(",") if "=" in p)
    try:
        t = int(parts.get("t", "0"))
    except ValueError:
        return False
    v1 = parts.get("v1", "")
    if not t or not v1 or abs(time.time() - t) > tolerance_sec:
        return False
    if isinstance(raw_body, str):
        raw_body = raw_body.encode()
    expected = hmac.new(secret.encode(), ("%d." % t).encode() + raw_body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, v1)
